fix: PreviewPhrase becomes invalid when adding second a phrase (M2-8199)

AlejandroCoronadoN opened the pull request:
### 📝 Description
The "Preview Phrase" button of the first phrase becomes disabled at the screen if second phrase is added

🔗 [Jira Ticket M2-8199](https://mindlogger.atlassian.net/jira/software/c/projects/M2/boards/4?assignee=712020%3A6322a990-2eaf-42f4-9fed-30af94baf20e&selectedIssue=M2-8199)


Changes include:

- Importing PhrasalTemplateResponseValuePhraseSchema on PhrasalTemplatePhrase to validate form input to display PhrasalPreview button.
- Deleted the isValid and isSubmmiting logic attached to the PhrasalPreview button and substitute it with the PhrasalTemplateResponseValuePhraseSchema validation. 

### 📸 Screenshots

<img width="1432" alt="image" src="https://github.com/user-attachments/assets/b30e537b-7267-42b7-95b6-2668153d7a7b">


### 🪤 Peer Testing

* Create an Applet
* Create at least one activity that is not a Phrase Builder.
* Create a new Phrase builder actyivity. 
* Complete the first phrase by filling the text, select the forst activitry you created as input for the second field and finally create a text for the last entry inside the Phrase. 
* At this point you should be able to click on the Preview Phrase button.
* Click on Add Phrase to insert a new Phrase
* The first Preview Phrase button should still by enabled and it should display the preview phrase. 

### ✏️ Notes
I making the validation using the PhrasalTemplateResponseValuePhraseSchema but and I'm passing field to the PhrasalTemplatePhrase to validate the full object. However I could have just validated the Array contained in this object that is already called by useFieldArray inside PhrasalTemplatePhrase (Avoiding passing field to the script in the first place)  Notice that we have an structure like fields.field.fields where fields.field is a PhrasalTemplateResponseValuePhraseSchema and fields.field.fields is a PhrasalTemplateResponseValuePhraseFieldSchema. 

`  const { fields, append, remove, swap } = useFieldArray({
    control: control as unknown as Control<{ values: TPhrasalTemplateField[] }>,
    name: phraseFieldsName as 'values',
  });`

AlejandroCoronadoN committed:
Importing PhrasalTemplateResponseValuePhraseSchema on PhrasalTemplatePhrase to validate form input to display PhrasalPreview button

AlejandroCoronadoN committed:
deleted unrequired code and added field as a dependency for useEffect

AlejandroCoronadoN committed:
Deleted unrequired comment

AlejandroCoronadoN committed:
Typing async function

farmerpaul commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -31,26 +31,38 @@ import {
 import { PhrasalTemplatePhraseProps } from './PhrasalTemplatePhrase.types';
 import { getFieldPlaceholders, getNewDefaultField } from '../PhrasalTemplate.utils';
 import { KEYWORDS } from '../PhrasalTemplateField/PhrasalTemplateField.const';
+import { PhrasalTemplateResponseValuePhraseSchema } from '../../../../../../pages/BuilderApplet/BuilderApplet.schema';
```
Nit: Long relative paths involving many consecutive `../` are generally discouraged. An absolute import in this case is easier to read:
```suggestion
import { PhrasalTemplateResponseValuePhraseSchema } from 'modules/Builder/pages/BuilderApplet/BuilderApplet.schema';
```

farmerpaul commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -31,26 +31,38 @@ import {
   const phraseFieldsName = `${name}.fields`;
-  const { control, setValue, getValues, formState } = useCustomFormContext();
-
+  const { control, setValue, getValues, formState, watch } = useCustomFormContext();
```
Nit: There's an ESLint warning about an unused variable, maybe when debugging your code earlier? Try to notice when you introduce ESLint warnings. (Tip: Whenever I create a PR, I usually review my PR's **Files changed** tab in GitHub to do a sanity check and catch debugging code that I meant to delete, because it happens a lot!)

It's true there are many existing ESLint warnings still in the codebase, but we are actively working to eliminate them as we encounter them. At the minimum, avoid introducing new ones. 😇🙏

farmerpaul commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -105,6 +117,16 @@ export const PhrasalTemplatePhrase = ({
 
+  const validatePhraseField = async (field: Record<'id', string>): Promise<boolean> => {
+    try {
+      await PhrasalTemplateResponseValuePhraseSchema.validate(field, { abortEarly: false });
```
Curious why you thought it necessary to include `abortEarly: false`? The default is `true` because it's more performant. I ran some tests locally and it seems to work fine using the default behaviour, but maybe I missed an edge case?
```suggestion
      await PhrasalTemplateResponseValuePhraseSchema.validate(field);
```

farmerpaul commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -31,26 +31,38 @@ import {
+    };
+
+    validateField(); // Call the async function
+  }, [field, formState, formState.isSubmitting]);
```
Nit: I think it's redundant to include both `formState` and `formState.isSubmitting` as deps here; changes to `formState.isSubmitting` is implied by changes to `formState`.
```suggestion
  }, [field, formState]);
```

mbanting replied:
Looks like this one still is unresolved and wasn't made as part of @AlejandroCoronadoN last [commit](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1977/commits/f6a19795dd1606b98a7498bfa4033abaabb9456c)

farmerpaul replied:
I've just taken care of this :+1:

farmerpaul reviewed (commented):
Great work @AlejandroCoronadoN! Nice and simple, straightforward solution. 💪🏻 The only comments I had really were nitpicks, nothing major, but I'll leave it you to respond before I approve.

AlejandroCoronadoN committed:
Fixed code for PR review, worng imports and unnecesary lines

farmerpaul commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -105,6 +117,16 @@ export const PhrasalTemplatePhrase = ({
+    } catch (error) {
+      return true;
+    }
+  };
```
The logic here seems to be backwards. A function that provides a validation result should be `true` if it's valid, `false` if invalid. I'll make a tweak to this to make the logic a little more intuitive.

farmerpaul commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -31,26 +32,37 @@ import {
+  field,
 }: PhrasalTemplatePhraseProps) => {
   const { t } = useTranslation('app');
+  // Use `useEffect` to trigger validation when `fields` or form state changes
```
This inline comment doesn't seem to be applicable to the line that follows it; I'll move it somewhere more meaningful.

mbanting commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -31,26 +32,37 @@ import {
+  useEffect(() => {
+    const validateField = async () => {
+      const result = await validatePhraseField(field);
+      setPreviewPhraseDisabled(result);
```
@farmerpaul this will also need to be changed to 
```suggestion
      setPreviewPhraseDisabled(!result);
```
after updating `validatePhraseField` although I'm sure you know this already :)

farmerpaul committed:
fix: PR tweaks

Moved `validatePhraseField` out of component render function since it's
static, and adjusted return value to be more intuitive. Made other minor
PR tweaks.

mbanting reviewed (approved):
Approving with the remaining changes made by @farmerpaul. Thank you @AlejandroCoronadoN and @farmerpaul !

farmerpaul committed:
fix: validate against dynamic field array

Previously, validation was happening only on the static list of field
array elements that were present when the page loaded. This fix ensures
elements added to the field array while editing the Phrase Builder item
are also accounted for during Preview button validation.
